### PR TITLE
[modelling] Change CostStack backend to an unordered_map type

### LINKED ISSUE #195: Replace vector of CostPtr in CostStack objects by a map of CostPtr
When creating a CostStack, the user must keep in mind the index of each Cost component they have defined, in order to be able to modify the Cost object later on. It would be more intuitive to use `std::map<String,CostPtr>` to retrieve each component, instead of using `std::vector<CostPtr>`.

### PR DESCRIPTION
**Pull request by ManifoldFR**
- Also adds a template getter for the component cost functions
- `addCost()` now has an overload taking a key (either a `std::size_t` or a string)
- `addCost()` returns a reference to the pair of cost function component and weight

Resolves #195

### CODE REVIEW

File: include/aligator/modelling/costs/sum-of-costs.hxx [hunk 1]
```diff
@@ -25,12 +25,19 @@ CostStackTpl<Scalar>::CostStackTpl(xyz::polymorphic<Manifold> space,
         ALIGATOR_RUNTIME_ERROR(msg);
       }
     }
+
+    for (std::size_t i = 0; i < comps.size(); i++) {
+      components_.emplace(
+          std::make_pair(i, std::make_pair(comps[i], weights[i])));
```
Comment (edantec) [inline include/aligator/modelling/costs/sum-of-costs.hxx:31]: Why do you emplace a std::pair here? Can't you just do `components_.emplace(i, std::make_pair(comps[i], weights[i])));`?

**Review by edantec — APPROVED** [review]
Apart from a small question, everything seems fine to me

File: include/aligator/modelling/costs/sum-of-costs.hxx (hunk 1 above)
Comment (ManifoldFR) [reply include/aligator/modelling/costs/sum-of-costs.hxx:31]: Nice catch, I think I was debugging and left it there

**Comment by ManifoldFR** [comment]
I'll force merge cos I'm the boss here

**Pull request merged**